File: 2023/day-2/part_one.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	sum := 0

	for _, game := range readLines("input.txt") {
		isPossible := true

		idAndResults := strings.Split(game, ": ")
		id, _ := strconv.Atoi(strings.Split(idAndResults[0], " ")[1])
		for _, result := range strings.Split(idAndResults[1], "; ") {
			cubes := map[string]int{}

			for _, cube := range strings.Split(result, ", ") {
				amountAndColour := strings.Split(cube, " ")
				amount, _ := strconv.Atoi(amountAndColour[0])
				cubes[amountAndColour[1]] = amount
			}

			if cubes["blue"] > 14 || cubes["green"] > 13 || cubes["red"] > 12 {
				isPossible = false
				break
			}
		}

		if isPossible {
			sum += id
		}
	}

	fmt.Println(sum)
	// 3059
}

func readLines(fileName string) (lines []string) {
	dir, _ := os.Getwd()
	filePath := fmt.Sprintf("%v/%v", dir, fileName)
	file, _ := os.Open(filePath)

	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return
}

File: 2023/day-2/part_one.rb
sum = File.readlines("#{__dir__}/input.txt").sum do |line|
  possible = line.split(": ").last.split("; ").all? do |result|
    cubes = result.split(", ").to_h { _1.split.reverse }
    cubes["blue"].to_i <= 14 && cubes["green"].to_i <= 13 && cubes["red"].to_i <= 12
  end

  possible ? line.match(/(?<id>\d+):/)[:id].to_i : 0
end

p sum
# 3059
